Main.hs

```diff
@@ -11,28 +11,32 @@ import qualified Data.List as L
 import Text.Printf (printf)
 
 multByIndex :: [Int] -> [Int]
-multByIndex = undefined
+multByIndex xs = zipWith (*) xs [0..]
+
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex xs = zipWith (^) xs [0..]
 
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference [] = error "The list is too short"
+productOfDifference [_] = error "The list is too short"
+productOfDifference xs = foldl (*) 1 (zipWith (flip (-)) (tail xs) xs)
```

Comment:
You have some code repetition in the first two cases; moreover, one of the matches is redundant. Notice, that you don't really need to check if your list has an even length, because of how `zipWith` is implemented.

Reply:
I decided to implement this with the labeled pattern matching (`@`) since I think that calling the function with a list of length one should throw an error, and there are no "pretty" ways to handle this in my current approach.

Reply:
> I think that calling the function with a list of length one should throw an error

Have you actually tried it? I'm sure it'll be OK.

Reply:
I mean that I *want* it to throw an error, since the list of length one is indeed, too short. In the new approach I don't have any repetitions and the function behaves as I want. I guess with the old approach I could've removed the case with the empty list, because `zipWith` throws an error for empty lists already. But I didn't like that the function would throw two different errors. The new approach is therefore better.

Reply:
> I mean that I _want_ it to throw an error. 

I see. Then your solution is the best. 

Throwing errors is not desirable in general, and it's even worse when the errors are unrelated to the actual problem.